### DISCUSSION TITLE
regex query help

### DISCUSSION THREAD

**Discussion by kskarthik** [blevesearch/bleve#1837] [Q&A] [answered]
Hello! I am trying to expose regex matching option for the user & i ran into a problem. 

It's not actually working. Code https://gitlab.com/kskarthik/go-ifsc/-/blob/master/cmd/root.go#L113

Am i missing anything?

![image](https://github.com/blevesearch/bleve/assets/11899221/f369a24e-e78f-49a4-aba9-4cc8c5f87e09)

When i use this pattern, the program panics

```
kar@earth:~/my/projects/go-ifsc$ go run main.go search -m regex ^ICIC*
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xc1a182]

goroutine 1 [running]:
ifsc/cmd.search({0xef7dc0?, 0xc000203e60}, {{0xc000203dd0, 0x1, 0x3}, {0x7ffe08fb2170, 0x5}, 0x64})
        /home/kar/my/projects/go-ifsc/cmd/root.go:148 +0x2c2
ifsc/cmd.SearchIFSC({{0xc000203dd0, 0x1, 0x3}, {0x7ffe08fb2170, 0x5}, 0x64})
        /home/kar/my/projects/go-ifsc/cmd/root.go:114 +0x433
ifsc/cmd.glob..func3(0x14dd120?, {0xc000203dd0?, 0x3?, 0x3?})
        /home/kar/my/projects/go-ifsc/cmd/search.go:34 +0x131
github.com/spf13/cobra.(*Command).execute(0x14dd120, {0xc000203d70, 0x3, 0x3})
        /home/kar/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0x14dcb60)
        /home/kar/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        /home/kar/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
ifsc/cmd.Execute()
        /home/kar/my/projects/go-ifsc/cmd/root.go:31 +0x25
main.main()
        /home/kar/my/projects/go-ifsc/main.go:8 +0x17
```

**Comment by metonymic-smokey** [comment, marked as answer]
@kskarthik thanks for reaching out.

I reproduced the panic locally with the same regex query. In this specific query, there is an error returned "zero width assertions not allowed" and hence, `results` was nil.
Adding error handling when dealing with returned errors in search() should fix this.

I added a log for the error:
```
result, err := index.Search(searchRequest)
fmt.Printf("error after query is %v \n", err)
```

This is the output after adding the log for the error:
```
aditiahuja@Aditis-MacBook-Pro go-ifsc % go run main.go search -m regex '^ICIC*'
error after query is zero width assertions not allowed 
```

Please let me know if you have any further questions, thanks!

**Reply by kskarthik** [reply]
Thank you very much! I have ignored to handle the err for search. My fault

**Comment by kskarthik** [comment]
@metonymic-smokey I am unable to do regex with this query https://blevesearch.com/docs/Query-String-Query/

could you help me out figuring out the right path to do so? https://gitlab.com/kskarthik/go-ifsc/-/blob/master/cmd/root.go#L113

Example:

I expect to get pattern matching for 

```sh
go run main.go search -m adv '/KARB0*/'
```

But matching results are not returned

Although this query works:

```sh
 go run main.go search -m adv '+delhi +axis'
```

**Comment by metonymic-smokey** [comment]
@kskarthik I don't think you can compare the two queries, since one is regex and one isn't.

For the regex query, maybe you could try using NewRegexpQuery() for the regex query ?

**Reply by kskarthik** [reply]
> @kskarthik I don't think you can compare the two queries, since one is regex and one isn't.

But, according to the docs, `bleve.NewQueryStringQuery()` supports regex queries too. 

> For the regex query, maybe you could try using NewRegexpQuery() for the regex query ?

I did try in the past. No use. But i will try once more.

**Comment by kskarthik** [comment]
@metonymic-smokey  where can i find more examples of using `bleve.NewRegexpQuery()` ? Didn't find here https://blevesearch.com/docs/Query/

**Comment by metonymic-smokey** [comment]
Hey @kskarthik 
Sorry for the delay. An example usage could be this test - https://github.com/blevesearch/bleve/blob/a569e452a8184a534a4a8694326b3d1dd31aaaec/search_test.go#L1480. 

One of the query types used is regexp.
https://github.com/blevesearch/bleve/blob/a569e452a8184a534a4a8694326b3d1dd31aaaec/search_test.go#L1574-L1575

Please let me know if you need any more info, thanks!
